Issue/1175/improve open api final 2

tschlat opened the pull request:
This is the final PR for new OpenAPI feature.

bbortt reviewed (changes requested):
I've reviewed and also refactored everything a bit. had the unique opportunity that my company offered me some time for this, so wanted to make the most out of it. thus the refactoring.

my changes are concluded in https://github.com/citrusframework/citrus/pull/1245, I will merge it into this MR shorty. I've made different commits for each module review.

the build is failing, currently, because of the `citrus-test-api-generator-core` module. I cannot reproduce it locally. I've also left a remark mentioning commit 67b8435 because of this.

I would definitely create a separate release for this implementation, but I like it. very well done @tschlat. so:
* create release from `main`
* merge MR
* create release from `main`

bbortt commented:
once done, the copyright headers should be re-applied to all `*.java` files.

bbortt commented:
found the last ci/cd bug. it was actually a case-sensitivity problem, which didn't occur on windows. we should expect to see a green build.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/OpenApiResourceLoader.java:
```diff
@@ -52,68 +61,108 @@ private OpenApiResourceLoader() {
+                con.setRequestProperty(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE);
+
+                int status = httpURLConnection.getResponseCode();
+                if (status > 299) {
```
I disagree:
Given the default behavior of HttpURLConnection, it automatically follows redirects, so the status code we receive will be from the final response, typically below 300. If we receive a 3xx status that is not automatically followed, it must be handled explicitly here. Since we are not handling such cases, receiving an unhandled 3xx status results in an unprocessed response, justifying the thrown exception.

bbortt replied:
that is true, it should follow automatically. didn't consider that.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+ */
+public class RandomConfiguration {
+
+    private static final String EMAIL_PATTERN = "[a-z]{5,15}\\.?[a-z]{5,15}\\@[a-z]{5,15}\\.[a-z]{2}";
```
This is just for generation. I remember to spent quite a while on this subject trying different patterns. But in the end i chose this one for simplicity.

tschlat replied:
I added a comment for clarification

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/OpenApiTestValidationDataGenerator.java:
```diff
@@ -0,0 +1,246 @@
+ *
+ * @author Christoph Deppisch
+ */
+public abstract class OpenApiTestValidationDataGenerator {
```
Seems to me like this should superfluous, as OpenAPI Validator is used for validation. But this will take a while for investigation.

tschlat replied:
Just checked it out. It is not necessary and therefore I have completely removed it. This means, that the validation on server side completely relies on the atlassian OpenAPI validator.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomElement.java:
```diff
@@ -0,0 +1,116 @@
+                this.putAll(randomObject);
+                return;
+            }
+            RandomElement.super.push(value);
```
Super is a LinkedHashMap. Anyhow, this is a little strange. I do not remember whether I did this with the purpose in mind to throw an exception. 

I propose to throw an IllegalArgumentException instead.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/validation/OpenApiMessageValidationContext.java:
```diff
@@ -0,0 +1,116 @@
+        public OpenApiMessageValidationContext.Builder expression(String path,
+            Object expectedValue) {
+            return new OpenApiMessageValidationContext.Builder().expression(path, expectedValue);
+        }
```
I guess that is why this method is not used.
Removed both.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/validation/OpenApiValidationContext.java:
```diff
@@ -0,0 +1,82 @@
+        return openApiInteractionValidator;
+    }
+
+    public synchronized SchemaValidator getSchemaValidator() {
```
This pattern is used throught citrus exactly this way, without "Default". Therefore, i would rather stick with it the way its is.

tschlat commented on core/citrus-base/src/main/java/org/citrusframework/functions/core/AdvancedRandomNumberFunction.java:
```diff
@@ -0,0 +1,227 @@
+
+        return decimalPlaces == 0 ?
+            format("%s", randomValue.longValue()) :
+            format(format("%%.%sf", decimalPlaces), randomValue.doubleValue());
```
yes i am sure. but to make it more readable i now use string concatenation for the inner format

tschlat commented on runtime/citrus-testng/src/main/java/org/citrusframework/testng/spring/TestNGCitrusSpringSupport.java:
```diff
@@ -163,6 +164,9 @@ protected void run(ITestResult testResult, Method method, List<TestLoader> metho
     public final void before() {
+        // We need to consider the possibility, that one test has meanwhile modified the current citrus instance,
+        // as there can be plenty of tests running between @BeforeSuite and the execution of an actual subclass of
+        // this support. The citrus instance may even have a mocked context.
```
Yes, we talked about it an no. It hasn't.

tschlat commented on test-api-generator/citrus-test-api-core/src/main/java/org/citrusframework/openapi/testapi/OpenApiParameterFormatter.java:
```diff
@@ -0,0 +1,226 @@
+        }
+
+        if (!rootNode.isObject()) {
+            throw new IllegalArgumentException("The provided JSON is not a valid JSON object.");
```
I changed it to:
"The provided string is not a valid JSON object."

tschlat commented on test-api-generator/citrus-test-api-generator-core/README.md:
```diff
@@ -0,0 +1,102 @@
+The code generator creates Java classes from an OpenAPI specification to facilitate API testing using the Citrus framework. 
+The Java classes represent Citrus Send and Receive ActionBuilders for each operation of the OpenAPI. Each builder provides
+setter for the actual operation parameters. In general a type-safe method is provided, that reflects the correct operation type.
+In addition a setter in string representation is provided to allow for citrus dynamic content setting using string expressions.
```
Work in progress

tschlat commented on test-api-generator/citrus-test-api-generator-core/README.md:
```diff
@@ -0,0 +1,102 @@
+## Validation
+
+It is known, that the used OpenAPI validator is not able to validate certain situations.
+E.g. certain array encoding situations related to object encoding 
```
Documentation is work in Progress

tschlat commented:
@ueberfuhr - as a user of the current OpenAPI feature, i would be interested in your feedback on this new feature. Especially in the context of OAS validation, which might be breaking, once we merge this MR (see MR description for details). I advise to check this class that tests all aspects of the generated API.

org.citrusframework.openapi.generator.GeneratedRestApiIT

tschlat commented on test-api-generator/citrus-test-api-core/src/main/java/org/citrusframework/openapi/testapi/ParameterStyle.java:
```diff
@@ -0,0 +1,25 @@
+    LABEL,
+    MATRIX,
+    FORM,
+    DEEPOBJECT
```
I noew remember why i didn't do this. deepObject is the wording in OpenApi spec. And this is directly translated in the mustache template:

```
ParameterStyle.{{#lambda.uppercase}}{{style}}
```

tschlat commented on test-api-generator/citrus-test-api-core/src/main/java/org/citrusframework/openapi/testapi/GeneratedApiRequest.java:
```diff
@@ -0,0 +1,45 @@
+ * Interface representing a generated API request corresponding to an operation in an OpenAPI specification.
+ * Provides methods to retrieve metadata about the request such as operation name, HTTP method, and path.
+ */
+public interface GeneratedApiRequest {
```
I renamed to the interface and now both Request and Response expose  method/path/operationname

tschlat commented on src/manual/connector-openapi.adoc:
```diff
@@ -309,3 +321,399 @@ Also, the server will verify the HTTP request method, the Content-Type header as
+| `version`                        | `citrus.test.api.generator.version`                           | Specifies the version of the API, may be null                    | (none)
+| `endpoint`                       | `citrus.test.api.generator.endpoint`                          | Specifies the endpoint of the test API                           | `applicationServiceClient`
+| `type`                           | `citrus.test.api.generator.type`                              | Specifies the type of the test API                               | `REST`, other option is `SOAP`
+| `useTags`                        | `citrus.test.api.generator.use.tags`                          | Specifies whether tags should be used by the generator           | `true`
```
Documentation is work in Progress

tschlat commented on test-api-generator/citrus-test-api-generator-core/README.md:
```diff
@@ -0,0 +1,102 @@
+
+Method names of non type-safe methods are prepended with a `$`. This is mainly to avoid conflicts for 
+
+### Structure of the Generated API Class
```
Documentation is work in Progress

tschlat commented on test-api-generator/citrus-test-api-generator-core/README.md:
```diff
@@ -0,0 +1,102 @@
+1. **Type-Safe Parameter Method**: This method accepts the required parameters directly.
+2. **String-Based Parameter Method**: This method accepts parameters as strings to allow for dynamic content.
+
+Method names of non type-safe methods are prepended with a `$`. This is mainly to avoid conflicts for 
```
Documentation is work in Progress

tschlat commented on src/manual/connector-openapi.adoc:
```diff
@@ -309,3 +321,399 @@ Also, the server will verify the HTTP request method, the Content-Type header as
+**  integration into XML editors via generated XSD
+***  schema validation
+***  auto completion
+*   integration into <<run-java,Citrus Java test cases>> via Java DSL [TODO #1161]
```
Documentation is work in Progress

tschlat commented on src/manual/connector-openapi.adoc:
```diff
@@ -309,3 +321,399 @@ Also, the server will verify the HTTP request method, the Content-Type header as
+[source,xml,indent=0,role="primary"]
+----
+<plugin>
+    <artifactId>citrus-test-api-generator-maven-plugin</artifactId>
```
I'm glad you liked it. If there's anything specific you think I should improve or if you have any suggestions, feel free to share.

bbortt replied:
that wasn't sarcastic at all! I just meant that I really like the feature/existence of this plugin 😉

bbortt commented on core/citrus-base/src/main/java/org/citrusframework/functions/core/AdvancedRandomNumberFunction.java:
```diff
@@ -134,7 +134,7 @@ private String getRandomNumber(int decimalPlaces,
         return decimalPlaces == 0 ?
             format("%s", randomValue.longValue()) :
-            format(format("%%.%sf", decimalPlaces), randomValue.doubleValue());
+            format("%." + decimalPlaces + "f", randomValue.doubleValue());
```
I think this is not the same functionality. or is it? I don't think so, because now you're printing the `decimalPlaces` literally, instead of repeated times.

tschlat replied:
It is the same and it is meant as such. Decimal places should be printed literally to form the correct format.

```java
System.out.println(format(format("%%.%sf", 4), 1.2345678));
System.out.println(format("%." + 4 + "f", 1.2345678));
```
1.2346
1.2346
```

```

bbortt replied:
thanks for clarification!

bbortt commented on pom.xml:
```diff
@@ -251,7 +251,8 @@
     <okhttp.version>4.12.0</okhttp.version>
-    <org.openapitools.version>7.5.0</org.openapitools.version>
+    <org.openapitools.version>7.9.0</org.openapitools.version>
+    <jackson-databind-nullable>0.2.6</jackson-databind-nullable>
```
```suggestion
    <jackson-databind-nullable.version>0.2.6</jackson-databind-nullable.version>
```

bbortt reviewed (changes requested):
reviewed and rebased it onto `origin/main`.

bbortt commented on endpoints/citrus-camel/src/test/java/org/citrusframework/camel/integration/CamelJBangIT.java:
```diff
@@ -58,6 +65,15 @@ public void runIntegrationWithSourceCodeIT() {
+        }
+
+        if (!TestUtils.isNetworkReachable()) {
+            throw new SkipException("Test skipped because network is not reachable. We are probably running behind a proxy and JBang download is not possible.");
```
removed this duplicate check.

bbortt commented on test-api-generator/citrus-test-api-generator-core/pom.xml:
```diff
@@ -284,6 +282,26 @@
                             </configOptions>
                         </configuration>
                     </execution>
+<!--                    <execution>-->
```
I've removed this.

bbortt reviewed (commented):
I've rebased this PR onto `origin/main`.

/lgtm

need to talk to @christophd about merging strategies.

tschlat commented on endpoints/citrus-http/src/main/java/org/citrusframework/http/client/HttpResponseErrorHandler.java:
```diff
@@ -46,7 +46,7 @@ protected void handleError(ClientHttpResponse response, HttpStatusCode statusCod
             throw new HttpErrorPropagatingException(response.getStatusCode(), response.getStatusText(), response.getHeaders(), getResponseBody(response), getCharset(response));
         } else if (errorHandlingStrategy.equals(ErrorHandlingStrategy.THROWS_EXCEPTION)) {
-            super.handleError(response);
+            super.handleError(response, statusCode, url, method);
```
@bbortt This fix should be added to Citrus Main

bbortt replied:
upstream `cherry-pick`: https://github.com/citrusframework/citrus/pull/1279.

tschlat commented:
Please review the last commit which also contains the documentation.

christophd commented:
Okay, this is a huge one and I postponed the review several times due to lack of time to have a look. Sorry for that. I will try to take a look over the weekend.

If possible and some stuff of it can be separated into a separate PR it would really help in the review process. Many thanks for the huge amount of work you put into this

bbortt commented:
@christophd it really is. I think helps knowing that we have an RC of this in place in production since beginning of 2025. which basically just means that "it works". splitting is difficult, I think. the only think I could think of is separate prs for:

* changes to existing code base
* newly added modules

tschlat commented:
> @christophd it really is. I think helps knowing that we have an RC of this in place in production since beginning of 2025. which basically just means that "it works". splitting is difficult, I think. the only think I could think of is separate prs for:
> 
> * changes to existing code base
> * newly added modules

I tried to avoid breaking the existing OpenAPI feature, but there are some potentially breaking changes:

- I replaced the control message validation with full-fledged schema validation, which may cause issues in existing tests.
- By default, the operation path now considers the OpenAPI base path. This aligns with the behavior of HTTP simulations in the simulator, ensuring consistency.

Additionally, there's a new feature in citrus-simulator that integrates with the OpenAPI feature, enabling scenario generation from OpenAPI specifications. I implemented a huge amount of tests for all this and we have beta-tested the new feature with several inhouse projects. So I am quite confident, that the current state has the required maturity to be moved to main. I spend a couple of hours in the last days for documentation. I am sure, that this would be a good starting point although I am also sure, that Timon has a couple of suggestions for the documentation.

tschlat commented:
After discussion with Timon I am rewriting the PR. I will provide smaller PR with specific meanings. So stay tuned before reviewing.

tschlat commented:
I took the time to perform a soft reset and carefully organized everything into individual commits as the OpenAPI feature evolved. I’m quite confident that the overall result is solid, but individual commits might have minor issues that only resolve fully when everything is merged together.

The first MR is:
https://github.com/citrusframework/citrus/pull/1311

There are 4 further branches named 
issue/1175/oas_2.....
issue/1175/oas_3.....
issue/1175/oas_4.....
issue/1175/oas_5.....

We will have to merge them one by one as review proceeds.

The last one contains all changes. So we might as well go with that one and review the code on commit level within one MR. It is up to you.

christophd commented:
@tschlat is this still a thing? or can we close the PR?
